File: problems/031-compile-critical-path/solutions/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const mod int64 = 1_000_000_007

func main() {
	in := bufio.NewReaderSize(os.Stdin, 1<<20)
	out := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer out.Flush()
	var n, m int
	fmt.Fscan(in, &n, &m)
	duration := make([]int64, n)
	for i := range duration {
		fmt.Fscan(in, &duration[i])
	}
	graph := make([][]int, n)
	indegree := make([]int, n)
	outdegree := make([]int, n)
	for i := 0; i < m; i++ {
		var u, v int
		fmt.Fscan(in, &u, &v)
		u--
		v--
		graph[u] = append(graph[u], v)
		indegree[v]++
		outdegree[u]++
	}
	best := make([]int64, n)
	ways := make([]int64, n)
	queue := make([]int, 0, n)
	for node := 0; node < n; node++ {
		if indegree[node] == 0 {
			best[node] = duration[node]
			ways[node] = 1
			queue = append(queue, node)
		}
	}
	for head := 0; head < len(queue); head++ {
		node := queue[head]
		for _, target := range graph[node] {
			candidate := best[node] + duration[target]
			if candidate > best[target] {
				best[target] = candidate
				ways[target] = ways[node]
			} else if candidate == best[target] {
				ways[target] = (ways[target] + ways[node]) % mod
			}
			indegree[target]--
			if indegree[target] == 0 {
				queue = append(queue, target)
			}
		}
	}
	var answer int64 = -1
	var count int64
	for node := 0; node < n; node++ {
		if outdegree[node] != 0 {
			continue
		}
		if best[node] > answer {
			answer = best[node]
			count = ways[node]
		} else if best[node] == answer {
			count = (count + ways[node]) % mod
		}
	}
	fmt.Fprintln(out, answer, count)
}

File: problems/031-compile-critical-path/solutions/cpp/main.cpp
#include <cstdint>
#include <deque>
#include <iostream>
#include <vector>

using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  constexpr int MOD = 1'000'000'007;
  int n, m;
  cin >> n >> m;
  vector<int64_t> duration(n);
  for (auto &value : duration)
    cin >> value;
  vector<vector<int>> graph(n);
  vector<int> indegree(n), outdegree(n);
  for (int i = 0; i < m; ++i) {
    int u, v;
    cin >> u >> v;
    --u;
    --v;
    graph[u].push_back(v);
    ++indegree[v];
    ++outdegree[u];
  }
  vector<int64_t> best(n);
  vector<int> ways(n);
  deque<int> queue;
  for (int node = 0; node < n; ++node) {
    if (indegree[node] == 0) {
      best[node] = duration[node];
      ways[node] = 1;
      queue.push_back(node);
    }
  }
  while (!queue.empty()) {
    const int node = queue.front();
    queue.pop_front();
    for (const int target : graph[node]) {
      const int64_t candidate = best[node] + duration[target];
      if (candidate > best[target]) {
        best[target] = candidate;
        ways[target] = ways[node];
      } else if (candidate == best[target]) {
        ways[target] += ways[node];
        if (ways[target] >= MOD)
          ways[target] -= MOD;
      }
      if (--indegree[target] == 0)
        queue.push_back(target);
    }
  }
  int64_t answer = -1;
  int count = 0;
  for (int node = 0; node < n; ++node) {
    if (outdegree[node] != 0)
      continue;
    if (best[node] > answer) {
      answer = best[node];
      count = ways[node];
    } else if (best[node] == answer) {
      count += ways[node];
      if (count >= MOD)
        count -= MOD;
    }
  }
  cout << answer << ' ' << count << '\n';
}
